src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlValidator.cs

```diff
@@ -0,0 +1,92 @@
+        /// <param name="utf8Base64UrlText">A span of UTF-8 text to validate.</param>
+        /// <param name="decodedLength">If the method returns true, the number of decoded bytes that will result from decoding the input UTF-8 text.</param>
+        /// <returns><see langword="true"/> if <paramref name="utf8Base64UrlText"/> contains a valid, decodable sequence of base-64 encoded data; otherwise, <see langword="false"/>.</returns>
+        /// <remarks>
+        /// where whitespace is defined as the characters ' ', '\t', '\r', or '\n' (as bytes).
+        /// </remarks>
+        public static bool IsValid(ReadOnlySpan<byte> utf8Base64UrlText, out int decodedLength) =>
+            Base64.IsValid<byte, Base64UrlByteValidatable>(utf8Base64UrlText, out decodedLength);
+
+        private const uint UrlEncodingPad = '%'; // allowed for url padding
+
+        private readonly struct Base64UrlCharValidatable : Base64.IBase64Validatable<char>
+        {
+            private static readonly SearchValues<char> s_validBase64UrlChars = SearchValues.Create("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_");
+
+            public static int IndexOfAnyExcept(ReadOnlySpan<char> span) => span.IndexOfAnyExcept(s_validBase64UrlChars);
+            public static bool IsWhiteSpace(char value) => Base64.IsWhiteSpace(value);
+            public static bool IsEncodingPad(char value) => value == Base64.EncodingPad || value == UrlEncodingPad;
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool ValidateAndDecodeLength(int length, int paddingCount, out int decodedLength) =>
+                Base64UrlByteValidatable.ValidateAndDecodeLength(length, paddingCount, out decodedLength);
+        }
+
+        private readonly struct Base64UrlByteValidatable : Base64.IBase64Validatable<byte>
+        {
+            private static readonly SearchValues<byte> s_validBase64UrlChars = SearchValues.Create(Base64UrlEncoderByte.EncodingMap);
+
+            public static int IndexOfAnyExcept(ReadOnlySpan<byte> span) => span.IndexOfAnyExcept(s_validBase64UrlChars);
+            public static bool IsWhiteSpace(byte value) => Base64.IsWhiteSpace(value);
+            public static bool IsEncodingPad(byte value) => value == Base64.EncodingPad || value == UrlEncodingPad;
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static bool ValidateAndDecodeLength(int length, int paddingCount, out int decodedLength)
+            {
+                // Padding is optional for Base64Url, so need to account remainder. If remainder is 1, then it's invalid.
+                int remainder = (int)((uint)length % 4);
+
+                if (remainder == 1)
+                {
+                    decodedLength = 0;
+                    return false;
+                }
+
+                decodedLength = (int)((uint)length / 4 * 3) + (remainder > 0 ? remainder - 1 : 0) - paddingCount;
```

Comment:
`Math.DivRem` for `remainder` and `decodedLength`?

For `Math.DivRem` make sure to use the `uint` overload, as it produces better machine code.